### PR TITLE
Pipeline Filtering

### PR DESCRIPTION
**Pull request by TortoiseHam**
also echo the fe_batch attribute in case that was already set (+1 squashed commit)

Squashed commits:
[42e0ec68] support FilteredData in transform (+2 squashed commits)
Squashed commits:
[64751ec4] support filtering in pipeline benchmark
[000d224d] add some comments (+5 squashed commits)
Squashed commits:
[03e17d97] catch typeerror specifically
[26bf388a] fix virtually all of the edge cases
[4d01d82b] solve speed and memory problems of current implementation. problems with edge conditions though
[20d8b827] try pre-batching the data in workers
[23549f9c] fix a test (+1 squashed commit)
Squashed commits:
[5c4c55ab] fix doc formatting (+1 squashed commit)
Squashed commits:
[44470a31] remove dev note (+1 squashed commit)
Squashed commits:
[8bfefcaf] some simplifications (+1 squashed commit)
Squashed commits:
[9c9ca3ed] undo apphub edits (+3 squashed commits)
Squashed commits:
[96300aea] more unit tests
[6417b3f9] more unit tests
[860e7453] fixes / docs (+1 squashed commit)
Squashed commits:
[89a04d8c] preliminary support for extenddataset (+1 squashed commit)
Squashed commits:
[aa9d5971] fix shuffling behavior (+14 squashed commits)
Squashed commits:
[fa47d1d4] cleanup
[3cf685f4] preliminary functionality for batch dataset
[e6caa3d7] revert to simpler batch dataset
[f4cbff10] Kind of ridiculous batch dataset solution, but might be working...
[6ded8879] solve batch shuffling for single proc
[0f580014] fix an issue where batch sampler would draw too many batches if final batch was supposed to be filtered away
[06f9e077] add tests for batch dataset
[eb4cae3e] get closer to working with batch dataset
[a36757c2] fix some batch behavior issues
[a6422f80] preliminary unit tests. Indicate that functionality is not yet correct when shuffle = True
[ae3eb9a1] preliminary attempt to handle batch datasets
[1cb83873] start trying to support droplast
[0e33c14e] allow with or without replacement filtering
[3c9444bd] preliminary test of filter functionality

### CODE REVIEW

File: fastestimator/dataset/batch_dataset.py [hunk 1]
```diff
@@ -69,8 +69,8 @@ def __init__(self,
         self.all_fe_datasets = False
         self._check_input()
         self.index_maps = []
-        self.reset_index_maps(0)
-        self.pad_value = None
+        self.child_reset_fns = [dataset.fe_reset_ds for dataset in self.datasets if hasattr(dataset, 'fe_reset_ds')]
+        self.fe_reset_ds(seed=0)
```
Comment (vbvg2008) [inline fastestimator/dataset/batch_dataset.py:73]: this will be overwritten later right?

**Review by vbvg2008 — COMMENTED** [review]
LGTM

**Review by vbvg2008 — APPROVED** [review]
